mergekit/metric_methods/all_metrics.py

```diff
@@ -0,0 +1,373 @@
+    v_proj = ungroup_tensor(v_proj, GQA_groups)
+
+    k_proj = restructure_tensor(k_proj, num_heads)
+    v_proj = restructure_tensor(v_proj, num_heads)
+    q_proj = restructure_tensor(q_proj, num_heads)
+    o_proj = restructure_tensor(o_proj.T, num_heads) # Output weights are split into heads by rows, not columns
+
+    return k_proj, v_proj, q_proj, o_proj
+
+def compute_histogram(tensor: torch.Tensor, n_bins: int) -> List[np.ndarray]:
+    bin_counts, bin_edges = np.histogram(tensor.numpy(), bins=n_bins)
+    bin_widths = np.diff(bin_edges)
+    return bin_counts, bin_edges, bin_widths
+
+def cossim_heatmap(A: torch.Tensor, B: torch.Tensor) -> torch.Tensor:
+    # Normalize the rows of both matrices
+    A_norm = A / A.norm(dim=1, keepdim=True)
+    B_norm = B / B.norm(dim=1, keepdim=True)
+    
+    # Compute the cosine similarity matrix
+    similarity_matrix = torch.mm(A_norm, B_norm.t())
+    
+    return similarity_matrix
+
+# Metric functions
+
+def SMAPE(
+    tensors: List[torch.Tensor], **_kwargs
+) -> Dict[str, Any]:
+    """Symmetric Mean Absolute Percentage Error (SMAPE)."""
+
+    numerator = torch.abs(tensors[0] - tensors[1])
+    denominator = (torch.abs(tensors[0]) + torch.abs(tensors[1]))
+    smape = torch.mean(torch.div(numerator, denominator), dim=1) 
+    
+    hist_info = compute_histogram(smape, 100)
+    return {
+        'SMAPE Histogram': {
+            'count': hist_info[0],
+            'edges': hist_info[1],
+            'widths': hist_info[2]
+        },
+        'SMAPE_mean': smape.mean().item(),
+        'SMAPE_std': smape.std().item()
+    }
+
+def cossim(
+    tensors: List[torch.Tensor], return_heatmap=False, **_kwargs
+) -> torch.Tensor:
+    """Cosine similarity"""
+    cossim = F.cosine_similarity(tensors[0], tensors[1], dim=1)
+
+    res = {}
+
+    if return_heatmap:
+        res.update({'Cossim Heatmap': cossim_heatmap(tensors[0], tensors[1])})
+
+    assert torch.isclose(cossim, cossim, atol=1e-6).all(), "NaNs in cosine similarity"
+    assert torch.isclose(cossim, cossim_heatmap(tensors[0], tensors[1]).diagonal(), atol=1e-4).all(), "Diagonal elements of cosine similarity matrix do not match"
+
+    hist_info = compute_histogram(cossim, 100)
+    res.update({
+        'cossim Histogram': {
+            'count': hist_info[0],
+            'edges': hist_info[1],
+            'widths': hist_info[2]
+        },
+        'cossim_mean': cossim.mean().item(),
+        'cossim_std': cossim.std().item()
+    })
+    return res
+
+def scale(
+    tensors: List[torch.Tensor], return_heatmap=False, **_kwargs
+) -> torch.Tensor:
+    """
+    Scale difference: ratio of absolute difference to average scale.
+    Complementary to cosine similarity, which measures the angle between two vectors and is invariant to scale.
+    """
+
+    norm_0 = torch.norm(tensors[0], dim=1)
+    norm_1 = torch.norm(tensors[1], dim=1)
+
+    res = {}
+
+    scale_diff = torch.abs(norm_0 - norm_1) / ((norm_0 + norm_1) / 2)
```

Comment:
Should we be doing something here to guard against dividing by zero?

Reply:
yep - norms are non-negative so adding small epsilon will be fine